features/steps/thens/alignment.py

```diff
@@ -172,3 +225,84 @@ def step_impl(context, inst):
                         yield ValidationOutcome(inst=inst, expected="same count of segments",
                                                 observed=observed_msg,
                                                 severity=OutcomeSeverity.ERROR)
+
+
+@gherkin_ifc.step(
+    'Each segment in the layout must have the same geometry type as its corresponding segment in the shape representation')
+def step_impl(context, inst):
+    # work back up the nesting tree to obtain the alignment
+    align = None
+    layout_entity_types = [
+        "IFCALIGNMENTHORIZONTAL",
+        "IFCALIGNMENTVERTICAL",
+        "IFCALIGNMENTCANT",
+    ]
+    for rel in inst.Nests:
+        layout = rel.RelatingObject
+        if layout.is_a().upper() in layout_entity_types:
+            for rel2 in layout.Nests:
+                align_ent = rel2.RelatingObject
+                if align_ent.is_a() == "IfcAlignment":
+                    align = ifc43x_alignment_validation.entities.Alignment().from_entity(align_ent)
+
+    if align is None:
+        msg = f"Error processing instance {str(inst)}. "
+        msg += "Expected an IfcAlignmentSegment nested 2 levels below an IfcAlignment."
+        raise ValueError(msg)
+
+    for idx, align_segment in enumerate(context.instances):
```

Comment:
Edit: I see you need the index. Maybe you have the equivalent of an `align.segments.index(inst)`?

Reply:
I'm using the index to retrieve the representation from the overall IfcCompositeCurve/IfcGradientCurve/IfcSegmentedReferenceCurve (depending on which layout is being processed).  This was a result of an original assumption that the length of IfcAlignmentSegments in the business logic would always be the same as the length of segments in the overall representation curve.  Since that was a faulty assumption due to the HELMERT transition type, I suspect that a different approach is warranted.

Reply:
This has been revised.  We now have a separate function to traverse back up the graph in order to obtain the activation entity.